compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,221 @@
+//! This pass tries to inline calls to brillig functions that have all constant arguments.
+use std::collections::{BTreeMap, HashSet};
+
+use acvm::acir::circuit::ErrorSelector;
+use noirc_frontend::{monomorphization::ast::InlineType, Type};
+
+use crate::{
+    errors::RuntimeError,
+    ssa::{
+        ir::{
+            function::{Function, FunctionId, RuntimeType},
+            instruction::{Instruction, InstructionId, TerminatorInstruction},
+            value::{Value, ValueId},
+        },
+        optimize_ssa_after_inline_const_brillig_calls, Ssa, SsaBuilder,
+    },
+};
+
+impl Ssa {
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn inline_const_brillig_calls(mut self, inliner_aggressiveness: i64) -> Self {
+        let error_selector_to_type = &self.error_selector_to_type;
+
+        // Collect all brillig functions so that later we can find them when processing a call instruction
+        let mut brillig_functions = BTreeMap::<FunctionId, Function>::new();
+        for (func_id, func) in &self.functions {
+            if let RuntimeType::Brillig(..) = func.runtime() {
+                let cloned_function = Function::clone_with_id(*func_id, func);
+                brillig_functions.insert(*func_id, cloned_function);
+            };
+        }
+
+        // Keep track of which brillig functions we couldn't completely inline: we'll remove the ones we could.
+        let mut brillig_functions_we_could_not_inline = HashSet::new();
+
+        for func in self.functions.values_mut() {
+            func.inline_const_brillig_calls(
+                &brillig_functions,
+                &mut brillig_functions_we_could_not_inline,
+                inliner_aggressiveness,
+                error_selector_to_type,
+            );
+        }
+
+        // Remove the brillig functions that are no longer called
+        for func_id in brillig_functions.keys() {
+            // We never want to remove the main function (it could be brillig if `--force-brillig` was given)
```

Comment:
It can even be Brillig on its own like [here](https://github.com/noir-lang/noir/blob/b74b4eaf7acf6becd0f98ab25d0d5ccde3affebe/test_programs/execution_success/brillig_cow/src/main.nr#L45)